src/strat_engine/blockdev.rs

```diff
@@ -55,87 +55,109 @@ impl BlockDev {
+                return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
+                                                          format!("{} is not a block device",
+                                                                  path.display()))));
+            }
 
-        if !force {
-            let mut buf = [0u8; 4096];
-            try!(f.read(&mut buf));
+            let dev = try!(Device::from_str(&path.to_string_lossy()));
+
+            let mut f = try!(OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(path)
+                .map_err(|_| {
+                    io::Error::new(ErrorKind::PermissionDenied,
+                                   format!("Could not open {}", path.display()))
+                }));
+
+            if !force {
+                let mut buf = [0u8; 4096];
+                try!(f.read(&mut buf));
+
+                if buf.iter().any(|x| *x != 0) {
+                    return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
+                                                              format!("First 4K of {} is not \
+                                                                       zeroed, need to use \
+                                                                       --force",
+                                                                      path.display()))));
+                }
+            }
 
-            if buf.iter().any(|x| *x != 0) {
+            let dev_size = try!(blkdev_size(&f));
+            if dev_size < MIN_DEV_SIZE {
                 return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
-                                                          format!("First 4K of {} is not \
-                                                                   zeroed, need to use --force",
-                                                                  path.display()))));
+                                                          format!("{} too small, {} minimum",
+                                                                  path.display(),
+                                                                  ByteSize::b(MIN_DEV_SIZE as usize)
+                                                                  .to_string(true)))));
             }
+            good_paths.push((path, dev, dev_size));
         }
 
-        let dev_size = try!(blkdev_size(&f));
-        if dev_size < MIN_DEV_SIZE {
-            return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
-                                                      format!("{} too small, {} minimum",
-                                                             path.display(),
-                                                             ByteSize::b(MIN_DEV_SIZE as usize)
-                                                                 .to_string(true)))));
+        let mut bds = Vec::new();
+        for (path, dev, dev_size) in good_paths {
```

Comment:
As pointed out, I think ```paths``` or ```paths.iter()``` is all that is necessary here, ```good_paths``` should be able to be removed entirely.